SPLAT-1743: vSphere - add host and vm based zonal

jcpowermac opened the pull request:
### Changes

- New feature gate `VSphereHostVMGroupZonal`
- Add `RegionType` and `ZoneType` fields to `VSpherePlatformFailureDomainSpec` to specify region and zone failure domain types.
- Add `VSphereFailureDomainAffinity` type that contains three fields required for vm-host zonal: `VMGroup`, `HostGroup` and `VMHostRule`.
- Add `VSphereFailureDomainAffinity` to `VSpherePlatformTopology`
- Create additional unit tests to cover the introduction of vm-host zonal provisioning types.
- Add `VMGroup` in a machine(s) workspace.

### Additional PRs

- https://github.com/openshift/enhancements/pull/1677
- https://github.com/openshift/installer/pull/8873
- https://github.com/openshift/client-go/pull/294
- https://github.com/openshift/library-go/pull/1782
- https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/325
- https://github.com/openshift/machine-api-operator/pull/1285

- https://github.com/openshift/api/pull/1999

jcpowermac commented:
(will re-write the commit message with appropriate details once this PR is closer to being ready)

/test build
/test e2e-aws-ovn

/test e2e-aws-ovn-techpreview

/test e2e-upgrade
/test e2e-upgrade-minor
/test images
/test integration
/test unit
/test verify
/test verify-client-go
/test verify-crd-schema
/test verify-deps

jcpowermac commented:
/test build
/test images
/test integration
/test unit
/test verify
/test verify-client-go
/test verify-crd-schema
/test verify-deps

jcpowermac commented:
/test build
/test images
/test integration
/test unit
/test verify
/test verify-client-go
/test verify-crd-schema
/test verify-deps

jcpowermac commented:
/test build
/test images
/test integration
/test unit
/test verify
/test verify-client-go
/test verify-crd-schema
/test verify-deps

jcpowermac commented:
/test build
/test images
/test integration
/test unit
/test verify
/test verify-client-go
/test verify-crd-schema
/test verify-deps

jcpowermac commented:
/test build
/test images
/test integration
/test unit
/test verify
/test verify-client-go
/test verify-crd-schema
/test verify-deps

jcpowermac commented:
/test unit

jcpowermac commented on tests/generator.go:
```diff
@@ -178,6 +179,8 @@ func generateOnCreateTable(onCreateTests []OnCreateTestSpec) {
 		expectedObj.SetName(gotObj.GetName())
 		expectedObj.SetNamespace(gotObj.GetNamespace())
 
+		fmt.Println(gotObj)
```
Will reset after review

jcpowermac commented on tests/suite_test.go:
```diff
@@ -37,7 +37,8 @@ func TestAPIs(t *testing.T) {
 	// this assumes a directory.  Because it does, we shall assume a path to the serialized featuregate manifests too.
-	suites, err = LoadTestSuiteSpecs(filepath.Join(".."))
+	//suites, err = LoadTestSuiteSpecs(filepath.Join(".."))
+	suites, err = LoadTestSuiteSpecs(filepath.Join("../config/v1/tests/infrastructures.config.openshift.io"))
```
will reset after review

rvanderp3 commented:
/lgtm

jcpowermac commented:
/assign @JoelSpeed

jcpowermac commented:
@JoelSpeed this pr and the enhancement https://github.com/openshift/enhancements/pull/1677
are ready for review when you have a chance

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1277,6 +1307,47 @@ type VSpherePlatformTopology struct {
+// +union
+type VSphereFailureDomainAffinity struct {
+	// +kubebuilder:validation:Optional
+	// +openshift:validation:FeatureGateAwareEnum:featureGate=VSphereHostVMGroupZonal,enum=HostGroup;ComputeCluster;Datacenter
```
Doesn't need to be a feature gate aware enum
As the parent field using this struct is gated, nothing in here needs to worry about being gated

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1277,6 +1307,47 @@ type VSpherePlatformTopology struct {
+	// +kubebuilder:validation:Optional
+	// +openshift:validation:FeatureGateAwareEnum:featureGate=VSphereHostVMGroupZonal,enum=HostGroup;ComputeCluster;Datacenter
+	// +unionDiscriminator
+	Type VSphereFailureDomainType `json:"type"`
```
You're missing the godoc for this field

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1277,6 +1307,47 @@ type VSpherePlatformTopology struct {
+	Type VSphereFailureDomainType `json:"type"`
+	// vmGroup is the name of the vm-host group of type virtual machine within vCenter for this failure domain.
+	// This field is required when the VSphereFailureDomain ZoneType is HostGroup
+	// +openshift:validation:featureGate=VSphereHostVMGroupZonal
```
Don't need to gate this field

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1277,6 +1307,47 @@ type VSpherePlatformTopology struct {
+	// +kubebuilder:validation:MaxLength=80
+	// +optional
+	// +unionMember,optional
+	VMGroup string `json:"vmGroup,omitempty"`
```
You only ever are going to need this string, any reason not to make this a struct so that you could add other fields in the future?

jcpowermac replied:
we have provided all the required fields that are in affinity configuration of a vCenter cluster

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1277,6 +1307,47 @@ type VSpherePlatformTopology struct {
+	// +openshift:validation:featureGate=VSphereHostVMGroupZonal
+	// +kubebuilder:validation:MaxLength=80
+	// +optional
+	// +unionMember,optional
```
You mean `+unionMember`, the `optional` here means that you can specify the type HostGroup without supplying this field, it's an optional bit of information to add to the context, but not required.

You've validated above that the field is required

```suggestion
	// +unionMember
```

jcpowermac replied:
When setting `RegionAffinity` only the type is required, all the string fields of `VSphereFailureDomainAffinity` are not required.

JoelSpeed replied:
Re-using this type between the region and zone affinities is likely to cause headache if they have validation requirements that are different to one another

JoelSpeed commented on machine/v1beta1/types_vsphereprovider.go:
```diff
@@ -190,6 +190,9 @@ type Workspace struct {
 	// ResourcePool is the resource pool in which VMs are created/located.
 	// +optional
 	ResourcePool string `gcfg:"resourcepool-path,omitempty" json:"resourcePool,omitempty"`
+	// VMGroup is the cluster vm group in which virtual machines will be added for vm host group based zonal.
+	// +optional
+	VMGroup string `gcfg:"vmGroup,omitempty" json:"vmGroup,omitempty"`
```
How does this interact with the infrastructure changes?

What is the implementation of this in CAPZ, how will we migrate this?

jcpowermac replied:
Its fine if you would rather see it as another pr. The machine needs the vmgroup for machine-api-operator to know which group to add a virtual machine into.

https://github.com/openshift/machine-api-operator/pull/1285/commits/50aa0d4b5357b3bda3e2738bfb6c60793c7c7dc1#diff-21a982202fdcdf4beffc4d00659798b54717940bf07f685006dbcb99d9dbf0ecR1143

jcpowermac replied:
CAPV vm-host zonal is already being used in the installer for this effort.

https://github.com/openshift/installer/pull/8873/commits/24d2fad365927f4f9728aed3acf0f018ea736b6f#diff-4f0946cfde9ce4a09f7a7fc9fb4a0b6b162f216e1ddc6ccbbfcac0965ce7e687R116-R119

jcpowermac commented on config/v1/types_infrastructure.go:
```diff
@@ -1277,6 +1307,47 @@ type VSpherePlatformTopology struct {
+	// +openshift:validation:FeatureGateAwareEnum:featureGate=VSphereHostVMGroupZonal,enum=HostGroup;ComputeCluster;Datacenter
+	// +unionDiscriminator
+	Type VSphereFailureDomainType `json:"type"`
+	// vmGroup is the name of the vm-host group of type virtual machine within vCenter for this failure domain.
```
I can add the godoc length. vmware is infuriating vague about even allowed lengths and there is no api documentation on allowed characters. We went through this in the original vsphere topology too. 


https://docs.vmware.com/en/VMware-vSphere/8.0/vsphere-web-services-803-programming-guide.pdf
On p30 they speak of using `%` to allow any character 
> The percent sign (%) is used as an escape character to embed special characters in strings.

JoelSpeed replied:
Even if they don't have a maximum length, we should impose one, even if it's arbitrary.

If the length is not limited, it's assumed to be `(3*1024^2)-2` for the purposes of any future CEL validation we might want to add

jcpowermac replied:
@JoelSpeed do you want length validation via CEL or via `kubebuilder:validation:MaxLength=80` ?

JoelSpeed replied:
Length validation must be via the kubebuilder annotation, as that's what dictates the cost estimate of the CEL rule :)

jcpowermac commented on features/features.go:
```diff
@@ -175,6 +175,13 @@ var (
 						enableIn(configv1.Default, configv1.DevPreviewNoUpgrade, configv1.TechPreviewNoUpgrade).
 						mustRegister()
 
+	FeatureGateVSphereHostVMGroupZonal = newFeatureGate("VSphereHostVMGroupZonal").
+						reportProblemsToJiraComponent("splat").
+						contactPerson("jcpowermac").
+						productScope(ocpSpecific).
+						enableIn(configv1.DevPreviewNoUpgrade, configv1.TechPreviewNoUpgrade).
+						mustRegister()
+
```
I just copied @vr4manta 's multiple vcenter example and changed, apparently that wasn't correct?

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1188,6 +1202,22 @@ type VSpherePlatformFailureDomainSpec struct {
+	// a vm-host group of type Virtual Machine and Host respectively. Is also
+	// contains the VMHostRule which is an affinity vm-host rule in vCenter.
+	// +openshift:validation:featureGate=VSphereHostVMGroupZonal
+	// +kubebuilder:validation:Optional
+	// +optional
```
Only need the latter of these two

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1188,6 +1202,22 @@ type VSpherePlatformFailureDomainSpec struct {
+	// a vm-host group of type Virtual Machine and Host respectively. Is also
+	// contains the VMHostRule which is an affinity vm-host rule in vCenter.
+	// +openshift:validation:featureGate=VSphereHostVMGroupZonal
+	// +kubebuilder:validation:Optional
+	// +optional
```
Only need the latter of these two

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1277,6 +1307,45 @@ type VSpherePlatformTopology struct {
+// This configuration within vCenter creates the required association between a failure domain, virtual machines
+// and ESXi hosts to create a vm-host based zone.
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'HostGroup' ?  has(self.hostGroup) : !has(self.hostGroup)",message="hostGroup is required when type is HostGroup, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'HostGroup' ?  has(self.vmGroup) : !has(self.vmGroup)",message="vmGroup is required when type is HostGroup, and forbidden otherwise"
```
Is this right? Normally there are different type values that map 1:1 with fields in the spec

If all of these fields are only required when type is hostGroup, they should all be in a hostGroup struct

If they are then optional for other values of type, you should make optional members that match up with those and include the fields for their type, with the correct validations

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1277,6 +1307,45 @@ type VSpherePlatformTopology struct {
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'HostGroup' ?  has(self.vmGroup) : !has(self.vmGroup)",message="vmGroup is required when type is HostGroup, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'HostGroup' ?  has(self.vmHostRule) : !has(self.vmHostRule)",message="vmHostRule is required when type is HostGroup, and forbidden otherwise"
+// +union
+type VSphereFailureDomainAffinity struct {
```
I'm not sure I'm following the validation requirements of these fields, for vmGroup, hostGroup and vmHostRule, could you fill the following table

| Type | HostGroup | ComputeCluster | DataCenter |
| --- | --- | --- | --- |
| vmGroup | | | |
| hostGroup | | | |
| vmHostRule | | | |

With values `R` for required, `O` for optional, or `N/A` if the field has no effect for that particular `Type` value

jcpowermac replied:
> I'm not sure I'm following the validation requirements of these fields, for vmGroup, hostGroup and vmHostRule, could you fill the following table

|Type|HostGroup| ComputeCluster |DataCenter
|---|---|---|---|
|vmGroup|R|N/A|N/A|
hostGroup |R|N/A|N/A|
vmHostRule|R|N/A|N/A|

> With values `R` for required, `O` for optional, or `N/A` if the field has no effect for that particular `Type` value

jcpowermac commented:
@JoelSpeed https://github.com/openshift/api/pull/1999/commits/2ee822a5c01f76da9639f182fac153ab09df51c1
changes from our meeting earlier today

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1188,6 +1208,19 @@ type VSpherePlatformFailureDomainSpec struct {
 
+	// regionAffinity holds the type of region
+	// +openshift:validation:featureGate=VSphereHostVMGroupZonal
+	// +kubebuilder:validation:Optional
```
Conversely to required, we prefer the plain version here
```suggestion
	// +optional
```

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1188,6 +1208,19 @@ type VSpherePlatformFailureDomainSpec struct {
 	// +kubebuilder:validation:Required
 	Zone string `json:"zone"`
 
+	// regionAffinity holds the type of region
```
Any way we can expand this explanation at all?

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1188,6 +1208,19 @@ type VSpherePlatformFailureDomainSpec struct {
+	// regionAffinity holds the type of region
+	// +openshift:validation:featureGate=VSphereHostVMGroupZonal
+	// +kubebuilder:validation:Optional
+	RegionAffinity VSphereFailureDomainRegionAffinity `json:"regionAffinity,omitempty"`
```
As this has `omitempty`, it needs to be a pointer else the json marshaller won't honour the omitempty

We only need to omit structs in configuration APIs when they have required fields, which I believe this one does

jcpowermac replied:
ack

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1277,6 +1310,86 @@ type VSpherePlatformTopology struct {
+// and the vm-host affinity rule that together creates an affinity configuration for vm-host based zonal.
+// This configuration within vCenter creates the required association between a failure domain, virtual machines
+// and ESXi hosts to create a vm-host based zone.
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'HostGroup' ?  has(self.hostGroup.hostGroup) : !has(self.hostGroup) || !has(self.hostGroup.hostGroup)",message="hostGroup is required when type is HostGroup, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'HostGroup' ?  has(self.hostGroup.vmGroup) :  !has(self.hostGroup) || !has(self.hostGroup.vmGroup)",message="vmGroup is required when type is HostGroup, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'HostGroup' ?  has(self.hostGroup.vmHostRule) : !has(self.hostGroup) || !has(self.hostGroup.vmHostRule)",message="vmHostRule is required when type is HostGroup, and forbidden otherwise"
```
I don't think these are correct anymore right?

jcpowermac replied:
I moved them above the VSphereFailureDomainZoneAffinity struct and then modified to include the scenario of a Region type: Datacenter, Zone type: ComputeCluster.

Based on the unit tests, they seem correct.

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1277,6 +1310,86 @@ type VSpherePlatformTopology struct {
+// +union
+type VSphereFailureDomainZoneAffinity struct {
+	// type is the string representation of the VSphereFailureDomainZoneType with available options of
+	// ComputeCluster and HostGroup.
```
Can we add
```
// When set to ComputeCluster, this means XYZ.
// When set to HostGroup, hostGroup must be configured with XYZ and this means XYZ.
```

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1277,6 +1310,86 @@ type VSpherePlatformTopology struct {
+	// +kubebuilder:validation:Enum:=HostGroup;ComputeCluster
+	// +kubebuilder:validation:MinLength=9
+	// +kubebuilder:validation:MaxLength=14
+	// +kubebuilder:validation:Type:=string
```
Shouldn't be needed, no harm in keeping

jcpowermac replied:
Without `// +kubebuilder:validation:Type:=string` I received an error that only strings are allowed to use min and maxlengths

JoelSpeed replied:
Interesting, must be something with it being picked up as an enum

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1277,6 +1310,86 @@ type VSpherePlatformTopology struct {
+	// +kubebuilder:validation:MaxLength=14
+	// +kubebuilder:validation:Type:=string
+	// +unionDiscriminator
+	Type VSphereFailureDomainZoneType `json:"type"`
```
Needs explicit required tag

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1277,6 +1310,86 @@ type VSpherePlatformTopology struct {
+	// contains the vmHostRule which is an affinity vm-host rule in vCenter.
+	// +unionMember,optional
+	HostGroup *VSphereFailureDomainHostGroup `json:"hostGroup,omitempty"`
+	// computeCluster currently is empty for future use.
+	// +unionMember,optional
+	ComputeCluster *VSphereFailureDomainComputeCluster `json:"computeCluster,omitempty"`
```
I would probably drop this until we actually need it, unless you want to force folks to add `computeCluster: {}`

I think adding this now, will make adding validation later harder

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1277,6 +1310,86 @@ type VSpherePlatformTopology struct {
+
+// VSphereFailureDomainRegionAffinity contains the region type and two currently
+// empty fields, datacenter and computeCluster for future use.
+type VSphereFailureDomainRegionAffinity struct {
```
Need union markers and CEL validations

jcpowermac replied:
Trying to think what CEL validations would be used here, the only thing I can think of is type, but there will be no members

jcpowermac replied:
Added CEL to check if type is `ComputeCluster` or `Datacenter`.

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1277,6 +1310,86 @@ type VSpherePlatformTopology struct {
+	// +unionDiscriminator
+	Type VSphereFailureDomainRegionType `json:"type"`
+
+	// datacenter currently is empty for future use.
+	// +unionMember,optional
+	Datacenter *VSphereFailureDomainDatacenter `json:"datacenter,omitempty"`
+	// computeCluster currently is empty for future use.
+	// +unionMember,optional
+	ComputeCluster *VSphereFailureDomainComputeCluster `json:"computeCluster,omitempty"`
```
I'd drop these for now, until explicitly needed

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1277,6 +1310,86 @@ type VSpherePlatformTopology struct {
+	// vmGroup is limited to 80 characters.
+	// This field is required when the VSphereFailureDomain ZoneType is HostGroup
+	// +kubebuilder:validation:MaxLength=80
+	// +kubebuilder:validation:Optional
```
I believe all three of these should be required now right?

jcpowermac replied:
Using the current CEL and required causes the unit tests to fail. based on the error required is checked first. 

Unless the unit tests should be changed since it isn't possible to have those fields empty?

JoelSpeed replied:
Perhaps the CEL is wrong, I'd expect CEL to require the parent struct, and all fields within then to be required, but not necessarily have CEL within them? 

Required strings probably want a minLength as well right?

jcpowermac commented:
/test minor-e2e-upgrade-minor

jcpowermac commented:
@JoelSpeed I pushed fe164067da0405e64fcbc4cd73d38f09b1a11a7e
with changes to CEL, required fields and MinLength. Included some additional unit tests.

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1277,6 +1315,82 @@ type VSpherePlatformTopology struct {
+	// Datacenter and ComputeCluster.
+	// When set to Datacenter, this means the vCenter Datacenter defined is the region.
+	// When set to ComputeCluster, this means the vCenter cluster defined is the region.
+	// +kubebuilder:validation:MinLength=9
+	// +kubebuilder:validation:MaxLength=14
```
Drop this and type string I think now

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1277,6 +1315,82 @@ type VSpherePlatformTopology struct {
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=80
+	// +kubebuilder:validation:Required
+	VMGroup string `json:"vmGroup"`
```
We found no way to limit the character set here right? Same question for the other 2 as well

jcpowermac replied:
Nope, previous thread:
https://github.com/openshift/api/pull/1999#discussion_r1801840403

JoelSpeed reviewed (commented):
In general, I think this is very close. Would like to see the CEL fixed up. You have the `hostGroup` struct as the union member now, so use that as the "required if and only if" type field. All the fields within it are required, so if it's present, become required

Would also like to see expansion on the errors that you've written out, they should ideally show the field that cause the error

Finally, are we certain that the freeform string fields we are adding do not have any restrictions on characters that are valid?

jcpowermac commented on config/v1/types_infrastructure.go:
```diff
@@ -1162,8 +1162,31 @@ type VSpherePlatformLoadBalancer struct {
+// VSpherePlatformFailureDomainSpec holds the region and zone failure domain and the vCenter topology of that failure domain.
+// +openshift:validation:FeatureGateAwareXValidation:featureGate=VSphereHostVMGroupZonal,rule="has(self.zoneAffinity) && self.zoneAffinity.type == 'HostGroup' ?  has(self.regionAffinity) && self.regionAffinity.type == 'ComputeCluster' : true",message="when zoneAffinity type is HostGroup, regionAffinity type must be ComputeCluster"
+// +openshift:validation:FeatureGateAwareXValidation:featureGate=VSphereHostVMGroupZonal,rule="has(self.zoneAffinity) && self.zoneAffinity.type == 'ComputeCluster' ?  has(self.regionAffinity) && self.regionAffinity.type == 'Datacenter' : true",message="when zoneAffinity type is ComputeCluster, regionAffinity type must be Datacenter"
+// +openshift:validation:FeatureGateAwareXValidation:featureGate=VSphereHostVMGroupZonal,rule="has(self.zoneAffinity) && self.zoneAffinity.type == 'HostGroup' ?  has(self.zoneAffinity) && has(self.zoneAffinity.hostGroup) : true",message="when zoneAffinity type is HostGroup, hostGroup must be defined"
```
This covers the supported region and zone affinity types 

Here is the comment w/ the table
https://github.com/openshift/api/pull/1999#discussion_r1811104637

jcpowermac commented:
@JoelSpeed rebased and pushed with the latest changes.

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1277,6 +1314,75 @@ type VSpherePlatformTopology struct {
+
+// VSphereFailureDomainRegionAffinity contains the region type which is the string representation of the
+// VSphereFailureDomainRegionType with available options of Datacenter and ComputeCluster.
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'Datacenter' || self.type == 'ComputeCluster'",message="regionAffinity type must be either Datacenter or ComputeCluster, and forbidden otherwise"
```
The enum validation you have should handle this, I don't think we need this in CEL

jcpowermac commented:
@JoelSpeed updated

JoelSpeed commented:
Verify looks like there's a need to re-run the update scripts? Diff in the generated CRDs

jcpowermac commented:
> Verify looks like there's a need to re-run the update scripts? Diff in the generated CRDs

It looks like I needed to rebase, just pushed

JoelSpeed commented:
/test verify-crd-schema

JoelSpeed commented:
/test minor-images
/test e2e-aws-ovn-hypershift
/test e2e-upgrade

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -1277,6 +1314,74 @@ type VSpherePlatformTopology struct {
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'HostGroup' ?  has(self.hostGroup) : !has(self.hostGroup)",message="hostGroup is required when type is HostGroup, and forbidden otherwise"
+// +union
+type VSphereFailureDomainZoneAffinity struct {
+	// type is the string representation of the VSphereFailureDomainZoneType with available options of
+	// ComputeCluster and HostGroup.
```
Does this focus too much on the implementation, rather the user facing side of this?
```suggestion
	// type determines the <something? Logical boundary?> for zones within this failure domain.
	// Available types are ComputeCluster and HostGroup.
```

JoelSpeed reviewed (commented):
One quick thought, otherwise this is good to go, LMK what you think of that question

JoelSpeed commented:
Huh, why did the integration test fail now? 🤔

jcpowermac commented:
> Huh, why did the integration test fail now? 🤔

I rebased, updated the sections and ran `make generate-with-container` 

what failed shouldn't have either, that is CEL to make sure the region and zone types are correctly paired

JoelSpeed commented:
> In tests the feature gate went from a string to a list

Ahh yes, we found certain tests needed the ability to specify that multiple gates were present, we made a breaking change to the test files, sorry :(

jcpowermac commented:
> > In tests the feature gate went from a string to a list
> 
> Ahh yes, we found certain tests needed the ability to specify that multiple gates were present, we made a breaking change to the test files, sorry :(

no worries, I am just glad it wasn't me :)

JoelSpeed commented:
/override ci/prow/verify-crd-schema

Incorrect in assuming that these fields are required and new, they have a new optional parents

/lgtm

JoelSpeed commented:
/override ci/prow/verify-crd-schema

JoelSpeed commented:
/override ci/prow/verify-crd-schema
